File: packages/repo-gc/src/shared/heuristics/dead-weight.ts
import { Finding, FindingKind, Severity, type FileData, type GraphData } from '../types';
import { estimateTokens } from '../token-estimate';

export interface DeadWeightOptions {
  /** File stem names that are always considered entry points (not orphaned). */
  entryPointStems?: string[];
  /** Regex patterns for config files that are never orphaned. */
  configPatterns?: RegExp[];
}

function resolveStem(relativePath: string): string {
  const parts = relativePath.split('/');
  const filename = parts[parts.length - 1];
  return filename.replace(/\.[^.]+$/, '');
}

export function analyzeDeadWeight(
  files: FileData[],
  graph: GraphData,
  idCounter: { value: number },
  options: DeadWeightOptions = {},
): Finding[] {
  const entryPointStems = options.entryPointStems ?? [];
  const configPatterns = options.configPatterns ?? [];

  // Build referenced-set from module declarations (Rust mod tree) + graph fan-in
  const referenced = new Set<string>();
  for (const f of files) {
    for (const decl of f.moduleDeclarations) {
      referenced.add(decl);
      // Add all prefixes so a reference to a.b.c also marks a and a.b
      const parts = decl.split('::');
      for (let i = 1; i < parts.length; i++) {
        referenced.add(parts.slice(0, i).join('::'));
      }
    }
  }

  const findings: Finding[] = [];

  for (const data of files) {
    // Entry points are never orphaned
    if (data.isEntryPoint) continue;

    // Package init files are never orphaned
    if (data.isPackageInit) continue;

    // Config files are never orphaned
    const stem = resolveStem(data.relativePath);
    if (entryPointStems.includes(stem)) continue;
    if (configPatterns.some((p) => p.test(data.relativePath))) continue;

    // Check if referenced via module declarations or import graph
    const fanIn = graph.fanIn.get(data.path) || 0;
    if (fanIn > 0) continue;
    if (referenced.has(data.modulePath)) continue;

    // Only flag if the file has meaningful content
    if (data.lineCount < 10) continue;

    let severity: Severity;
    if (data.lineCount >= 1000) severity = Severity.Critical;
    else if (data.lineCount >= 500) severity = Severity.High;
    else severity = Severity.Medium;

    const tokens = estimateTokens(data.sizeBytes);

    findings.push({
      id: `dw-${String(idCounter.value++).padStart(3, '0')}`,
      kind: FindingKind.DeadWeight,
      severity,
      confidence: 0.6,
      path: data.relativePath,
      summary: '',
      reasons: [],
      evidence: [
        `module_path: ${data.modulePath}`,
        `line_count: ${data.lineCount}`,
        `stem: ${stem}`,
        `estimated_tokens: ${tokens}`,
      ],
      suggested_next_step: '',
      estimated_tokens: tokens,
    });
  }

  return findings;
}

File: packages/repo-gc-python/python/repo_gc_python/heuristics/dead_weight.py
"""Unreferenced module detection — mirrors Rust's dead_weight.rs."""

from pathlib import Path

from ..discovery import PythonFile
from ..graph import ImportGraph
from ..parsing import FileInfo
from ..types import Finding, FindingKind, Severity
from .context_bombs import estimate_tokens


def analyze_orphaned_files(
    pfiles: list[PythonFile],
    infos: list[FileInfo],
    graph: ImportGraph,
    counter: int,
) -> list[Finding]:
    # Build a set of all module paths referenced by any file's imports.
    # Include all prefixes so "mypackage.foo.bar" also marks "mypackage.foo".
    referenced: set[str] = set()

    for info in infos:
        for mod_path in info.imported_modules:
            parts = mod_path.split(".")
            for length in range(1, len(parts) + 1):
                referenced.add(".".join(parts[:length]))

    # Build path -> info lookup
    info_by_path: dict[Path, FileInfo] = {}
    for info in infos:
        info_by_path[info.path] = info

    findings: list[Finding] = []

    for pfile in pfiles:
        # Entry points are never orphaned
        info = info_by_path.get(pfile.path)
        if info is not None and (info.is_entry_point or info.is_init_file):
            continue

        # Fallback: check the stem
        stem = pfile.path.stem
        if stem in ("__init__", "__main__", "conftest"):
            continue

        module_path = info.module_path if info is not None else f"{pfile.package_name}.{stem}"

        # Check if this module or any prefix is referenced
        parts = module_path.split(".")
        is_referenced = module_path in referenced or any(
            ".".join(parts[:length]) in referenced for length in range(1, len(parts))
        )

        if not is_referenced:
            # Skip files with <10 lines (empty/near-empty files are not meaningful dead weight)
            if pfile.line_count < 10:
                continue

            # Check import-graph fan-in: if any other file imports this one, it is not orphaned
            if graph.fan_in.get(pfile.path, 0) > 0:
                continue

            severity = (
                Severity.Critical
                if pfile.line_count >= 1000
                else Severity.High
                if pfile.line_count >= 500
                else Severity.Medium
            )

            counter += 1
            findings.append(
                Finding(
                    id=f"dw-{counter:03d}",
                    kind=FindingKind.DeadWeight,
                    severity=severity,
                    confidence=0.6,
                    path=str(pfile.relative_path),
                    summary="",
                    reasons=[],
                    evidence=[
                        f"module_path: {module_path}",
                        f"line_count: {pfile.line_count}",
                        f"stem: {stem}",
                        f"estimated_tokens: {estimate_tokens(pfile.size_bytes)}",
                    ],
                    suggested_next_step="",
                    estimated_tokens=estimate_tokens(pfile.size_bytes),
                )
            )

    return findings
